pkg/controller/metrics/metrics.go

```diff
@@ -339,6 +396,51 @@ func (mc *Calculator) Start(ctx context.Context) error {
 	return nil
 }
 
+// registerOptionalMetrics registers the metrics, and stores their configs in the corresponding maps
+func registerOptionalMetrics(mConfig *metricsconfig.MetricsConfig) {
+	MapMetricToDuration = make(map[*prometheus.HistogramVec]time.Duration)
+	for _, metric := range mConfig.MetricsWithDuration {
+		switch metric.Name {
+		case metricsconfig.CurrentStopping:
+			metrics.Registry.MustRegister(MetricStoppingClustersSeconds)
+			MapMetricToDuration[MetricStoppingClustersSeconds] = metric.Duration.Duration
+		case metricsconfig.CurrentResuming:
+			metrics.Registry.MustRegister(MetricResumingClustersSeconds)
+			MapMetricToDuration[MetricResumingClustersSeconds] = metric.Duration.Duration
+		case metricsconfig.CurrentWaitingForCO:
+			metrics.Registry.MustRegister(MetricWaitingForCOClustersSeconds)
+			MapMetricToDuration[MetricWaitingForCOClustersSeconds] = metric.Duration.Duration
+		case metricsconfig.CumulativeHibernated:
+			metrics.Registry.MustRegister(MetricClusterHibernationTransitionSeconds)
+			MapMetricToDuration[MetricClusterHibernationTransitionSeconds] = metric.Duration.Duration
+		case metricsconfig.CumulativeResumed:
+			metrics.Registry.MustRegister(MetricClusterReadyTransitionSeconds)
+			MapMetricToDuration[MetricClusterReadyTransitionSeconds] = metric.Duration.Duration
+		}
+	}
+}
+
+func existsInDurationMetricsMap(metric *prometheus.HistogramVec) bool {
+	if MapMetricToDuration != nil {
+		_, ok := MapMetricToDuration[metric]
+		return ok
+	}
+	return false
+}
+
+// ShouldLogDurationMetric decides whether the corresponding duration metric should be logged. It first checks if that
+// optional metric has been opted into for logging, and the duration has crossed the threshold set
+func ShouldLogDurationMetric(metricToLog *prometheus.HistogramVec, timeToLog float64) bool {
+	if MapMetricToDuration != nil {
+		for metric, duration := range MapMetricToDuration {
+			if metric == metricToLog && duration.Seconds() <= timeToLog {
+				return true
+			}
+		}
+	}
+	return false
```

Comment:
- Shouldn't need the nil check as `registerOptionalMetrics` does a `make()`.
- Should be able to look up (O(1)) rather than iterate (O(N)).
```suggestion
	duration, ok := MapMetricToDuration[metricToLog]
	return ok && duration.Seconds() <= timeToLog
```